src/Materials/Textures/Packer/packer.ts

```diff
@@ -0,0 +1,783 @@
+    * @returns Promise<void>
+    */
+    public processAsync(): Promise<void> {
+            return new Promise ((resolve, reject) => {
+                try {
+                    if (this.meshes.length === 0) {
+                        //Must be a JSON load!
+                        resolve();
+                        return;
+                    }
+                    let done = 0;
+                    const doneCheck = (mat: Material) => {
+                        done++;
+                        //Check Status of all Textures on all meshes, till they are ready.
+                        if (this.options.map) {
+                            for (let j = 0; j < this.options.map.length; j++) {
+                                let index: string = this.options.map[j];
+                                let t: (Texture | DynamicTexture) = (mat as any)[index];
+
+                                if (t !== null) {
+                                    if (!(this.sets as any)[this.options.map[j]]) {
+                                        (this.sets as any)[this.options.map[j]] = true;
+                                    }
+
+                                    this._expecting++;
+                                }
+                            }
+
+                            if (done === this.meshes.length) {
+                                this._createFrames(resolve);
+                            }
+                        }
+                    };
+
+                    for (let i = 0; i < this.meshes.length; i++) {
+
+                        let mesh = this.meshes[i];
+                        let material: Nullable< Material > = mesh.material;
+
+                        if (!material) {
+                            done++;
+                            if (done === this.meshes.length) {
+                                return this._createFrames(resolve);
+                            }
+                            continue;
+                        }
+
+                        material.forceCompilationAsync(mesh).then(() => {
+                            doneCheck((material as Material));
+                        });
+                    }
+                } catch (e) {
+                    return reject(e);
+                }
+            });
+    }
+
+    /**
+    * Disposes all textures associated with this packer
+    */
+    public dispose(): void {
+        let sKeys = Object.keys(this.sets);
+        for (let i = 0; i < sKeys.length; i++) {
+            let channel = sKeys[i];
+            (this.sets as any)[channel].dispose();
+        }
+    }
+
+    /**
+    * Starts the download process for all the channels converting them to base64 data and embedding it all in a JSON file.
+    * @param imageType is the image type to use.
+    * @param quality of the image if downloading as jpeg, Ranges from >0 to 1.
+    */
+    public download(imageType: string = 'png', quality: number = 1): void {

```

Comment:
This function should return the download promise so user can take action when download is done

Reply:
will do

Reply:
actually thinking about this one, I want control of what happens to the then.

Otherwise I would have to expect the user to pack it into the JSON format that I have setup.

Reply:
Maybe not even use a promise, but go ahead and use a setTimeout hack?

Reply:
Well this works for now..Will see if more is required when people get their hands on it :)